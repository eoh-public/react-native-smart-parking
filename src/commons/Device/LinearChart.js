import React, { memo, useState, useEffect, useCallback } from 'react';
import { StyleSheet, View } from 'react-native';
import HighchartsReactNative from '@highcharts/highcharts-react-native';
import moment from 'moment';
import { Colors } from '../../configs';

const convertData = (data = []) => {
  let arr = [];
  for (let i = 0; i < data.length; i++) {
    arr.push([moment(data[i].x).valueOf(), data[i].y]);
  }
  return arr;
};
const chartOptions = {
  chart: {},
  credits: {
    enabled: false,
  },
  rangeSelector: {
    selected: 1,
  },
  yAxis: {
    title: {
      text: '',
    },
  },
  title: {
    text: '',
  },

  series: [
    {
      name: 'AQI',
      data: [],
      type: 'spline',
      tooltip: {
        valueDecimals: 2,
      },
    },
  ],
  time: {
    timezoneOffset: -7 * 60,
  },
  xAxis: {
    type: 'datetime',
    title: {
      text: '',
    },
    labels: {
      formatter: function (value) {
        const time = new Date(this.value);
        let date = time.getDate();
        let month = time.getMonth();
        let hours = time.getHours();
        let minutes = time.getMinutes();
        if (hours < 10) {
          hours = '0' + hours;
        }
        if (minutes < 10) {
          minutes = '0' + minutes;
        }
        if (hours === '00' && minutes === '00') {
          if (date < 10) {
            date = '0' + date;
          }
          if (month < 10) {
            month = '0' + month;
          }
          return `${date}.${month} ${hours}:${minutes}`;
        } else {
          return `${hours}:${minutes}`;
        }
      },
    },
    minRange: 3600 * 1000,
  },
  plotOptions: {
    series: {
      point: {
        events: {
          click: () => {
            let series = this.series.chart.series;
            const seriesIndex = this.series.index;
            for (let i = 0; i < series.length; i++) {
              if (series[i].index !== seriesIndex) {
                series[i].visible ? series[i].hide() : series[i].show();
              }
            }
            return false;
          },
        },
      },
      events: {
        legendItemClick: (event) => {
          if (!this.visible) {
            return true;
          }
          let seriesIndex = this.index;
          let series = this.chart.series;
          for (let i = 0; i < series.length; i++) {
            if (series[i].index !== seriesIndex) {
              series[i].visible ? series[i].hide() : series[i].show();
            }
          }
          return false;
        },
      },
      showInNavigator: true,
    },
  },
};
const LinearChart = memo(({ datas }) => {
  const [options, setOption] = useState(chartOptions);
  const getSeries = useCallback(() => {
    let series = [];
    for (let i = 0; i < datas.length; i++) {
      const arr = convertData(datas[i].data);

      series.push({
        name: datas[i].title,
        data: arr,
        type: 'spline',
        color: datas[i].color,
      });
    }
    return series;
  }, [datas]);

  const updateChart = useCallback(() => {
    setOption({
      ...chartOptions,
      series: getSeries(),
    });
  }, [getSeries]);

  useEffect(() => {
    updateChart();
  }, [updateChart]);

  return (
    <View style={styles.container}>
      <HighchartsReactNative styles={styles.chartStyle} options={options} />
    </View>
  );
});

export default LinearChart;

const styles = StyleSheet.create({
  container: {
    marginTop: 24,
    backgroundColor: Colors.White,
    justifyContent: 'center',
    flex: 1,
    marginRight: 24,
  },
  chartStyle: {
    backgroundColor: Colors.White,
    flex: 1,
  },
});
