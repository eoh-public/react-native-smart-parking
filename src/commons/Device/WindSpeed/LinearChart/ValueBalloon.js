import React, { memo } from 'react';
import { G, Rect } from 'react-native-svg';

const ValueBalloon = memo((props) => {
  const { x, y } = props;
  const transform = `translate(${x - 24},${y - 54})`;

  return (
    <G transform={transform}>
      <Rect
        x={23.5752}
        y={29.3477}
        width={9.03012}
        height={9.03012}
        rx={1.12876}
        transform="rotate(45 23.5752 29.3477)"
        fill="#262626"
      />
      <Rect width={47.1506} height={36.1205} rx={3.38629} fill="#262626" />
    </G>
  );
});

export default ValueBalloon;
