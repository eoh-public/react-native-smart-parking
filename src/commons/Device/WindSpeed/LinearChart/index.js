import React, { memo } from 'react';
import Svg, { G } from 'react-native-svg';
import {
  VictoryLine,
  VictoryAxis,
  VictoryScatter,
  VictoryTooltip,
} from 'victory-native';
import { t } from 'i18n-js';

import { Colors } from '../../../../configs';
import ChartLoading from '../../../ChartLoading';
import ValueBalloon from '../../../Device/WindSpeed/LinearChart/ValueBalloon';
import { getMaxValueIndex } from '../../../../utils/chartHelper/getMaxValueIndex';
import getTickValues from '../../../../utils/dateHelper/getTickValues';

const LinearChart = memo(
  ({
    startTime,
    endTime,
    datasShow,
    configuration,
    datasColor,
    chartOptions,
  }) => {
    const datas = chartOptions.showAll
      ? datasShow
      : datasShow.filter((item, index) => index === chartOptions.index);
    if (!datasShow[0].data) {
      return <ChartLoading />;
    }
    if (!datasShow[0].data.length) {
      return <ChartLoading message={t('no_data')} />;
    }
    const chartTick = getTickValues(datas[0], configuration).stringArr;
    const maxValue = getMaxValueIndex(datas[0].data);

    return (
      <Svg width={'100%'} height={300} style>
        <G x={-12} y={-12}>
          <VictoryAxis
            dependentAxis
            domain={[0, maxValue.max.y * 1.2]}
            offsetX={50}
            orientation="left"
            standalone={false}
            style={styleVerticalAxis}
          />
          <VictoryAxis
            crossAxis
            scale="time"
            standalone={false}
            style={styleHorizontalAxis}
            tickValues={chartTick}
          />
          {datas.map((item, index) => {
            const tickValues = getTickValues(item).arr;
            const maxValueLine = getMaxValueIndex(item.data);

            return (
              <G key={index.toString()}>
                <VictoryLine
                  data={item.data}
                  domain={{
                    x: [tickValues[0], tickValues[tickValues.length - 1]],
                    y: [0, maxValueLine.max.y * 1.2],
                  }}
                  style={{
                    data: { stroke: item.color },
                  }}
                  interpolation="monotoneX"
                  scale={{ x: 'time', y: 'linear' }}
                  standalone={false}
                />
              </G>
            );
          })}
          {datas.map((item, index) => {
            const scatterValues = getTickValues(item).stringDataArr;
            const maxValueLine = getMaxValueIndex(item.data);
            return (
              <G>
                <VictoryScatter
                  standalone={false}
                  data={scatterValues}
                  labels={(datum) => {
                    if (item.measure === 'pH') {
                      return `${item.measure} ${datum.y.toString()}`;
                    }
                    return `${datum.y.toString()} ${item.measure}`;
                  }}
                  labelComponent={
                    <VictoryTooltip flyoutComponent={<ValueBalloon />} />
                  }
                  domain={{ y: [0, maxValueLine.max.y * 1.2] }}
                  samples={25}
                  size={5}
                  style={{ ...styleScatter, data: { fill: item.color } }}
                />
              </G>
            );
          })}
        </G>
      </Svg>
    );
  }
);

export default LinearChart;

const styleVerticalAxis = {
  axis: { stroke: Colors.Gray4, strokeWidth: 1 },
  ticks: {
    size: 5,
    stroke: 'black',
    strokeWidth: 0,
  },
  tickLabels: {
    fill: Colors.Gray8,
    fontFamily: 'SFProDisplay-Regular',
    fontSize: 10,
  },
  grid: {
    fill: 'none',
    stroke: Colors.Gray4,
    strokeDasharray: '5,5',
    pointerEvents: 'painted',
  },
};

const styleHorizontalAxis = {
  axis: { stroke: 'black', strokeWidth: 0 },
  ticks: {
    size: 10,
    stroke: 'black',
    strokeWidth: 0,
  },
  tickLabels: {
    fill: Colors.Gray8,
    fontFamily: 'SFProDisplay-Regular',
    fontSize: 10,
  },
};

const styleScatter = {
  labels: {
    fill: Colors.White,
    padding: 14,
    fontFamily: 'SFProDisplay-Regular',
    fontSize: 12,
  },
};
