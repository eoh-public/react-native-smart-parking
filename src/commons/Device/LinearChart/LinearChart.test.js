import React from 'react';
import renderer, { act } from 'react-test-renderer';
import LinearChart from '../LinearChart';

describe('Test LinearChart', () => {
  let tree;
  test('render LinearChart', () => {
    const datas = [
      {
        id: 1,
        title: 'title',
        color: 'red',
        data: [
          [1, 2],
          [2, 3],
        ],
      },
    ];
    act(() => {
      tree = renderer.create(<LinearChart datas={datas} />);
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
