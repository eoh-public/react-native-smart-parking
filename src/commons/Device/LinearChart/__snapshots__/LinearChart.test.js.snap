// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test LinearChart render LinearChart 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "justifyContent": "center",
      "marginRight": 24,
      "marginTop": 24,
    }
  }
>
  <View />
</View>
`;
